src/Aurora.Shared/Models/ValueOrNull.cs

```diff
@@ -79,12 +93,28 @@ public static ValueOrNull<T> CreateNull(string? nullMessage = null)
 
 public static class ValueOrNullExtensions
 {
-    public static T WithDefault<T>(this ValueOrNull<T> valueOrNull, T defaultValue, Action<string>? onError = null)
-    {
-        return valueOrNull.Resolve<T>(x => x, msg =>
+    public static ValueOrNull<U> PipeValue<T, U>(this ValueOrNull<T> valueOrNull, Func<T, U> transform) =>
+        valueOrNull.Resolve(
+            value => transform(value).ToResult(), 
+            errorMessage => errorMessage.ToErrorResult<U>()
+            );
+
+    public static ValueOrNull<U> PipeValue<T, U>(this ValueOrNull<T> valueOrNull, Func<T, ValueOrNull<U>> transform) =>
+        valueOrNull.Resolve(
+            value => transform(value),
+            errorMessage => errorMessage.ToErrorResult<U>()
```

Comment:
may be use not one error but list of errors, or add both, also you can add Error abstract class and Public and Private errors, such that Public is for user interface and Private just system-internal, this error and errors could get Error and ErrorList accordingly

Reply:
That is a good suggestion, but we currently do not expose any errors to the end-user, so that would not be valuable at the moment